pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java

```diff
@@ -129,7 +129,7 @@ public class Consumer {
 
     private static final double avgPercent = 0.9;
     private boolean preciseDispatcherFlowControl;
-    private PositionImpl readPositionWhenJoining;
+    private volatile PositionImpl readPositionWhenJoining;
```

Comment:
The `PositionImpl` class has two mutable `long` fields. Given that it is only ever serialized to a `String`, we might avoid some edge cases by making this object a `volatile String`.

Reply:
that's true. the design of PositionImpl is not optimal since it's not an immutable class. However, in this case, the instance that is used is an "effectively immutable" instance.